src/base/memory_monitor_sr.hpp

```diff
@@ -0,0 +1,234 @@
+} MMON_STAT_ID;
+
+/* APIs */
+int mmon_add_stat (THREAD_ENTRY *thread_p, MMON_STAT_ID stat_id, uint64_t size);
+int mmon_sub_stat (THREAD_ENTRY *thread_p, MMON_STAT_ID stat_id, uint64_t size);
+int mmon_move_stat (THREAD_ENTRY *thread_p, MMON_STAT_ID src, MMON_STAT_ID dest, uint64_t size);
+int mmon_resize_stat (THREAD_ENTRY *thread_p, MMON_STAT_ID stat_id, uint64_t old_size, uint64_t new_size);
+
+namespace cubperf
+{
+  typedef struct mmon_comp_info
+  {
+    MMON_STAT_ID id;
+    const char *comp_name;
+    const char *subcomp_name;
+  } MMON_COMP_INFO;
+
+  typedef struct mmon_stat
+  {
+    std::atomic<uint64_t> init_stat;
+    std::atomic<uint64_t> cur_stat;
+    std::atomic<uint64_t> peak_stat;
+    std::atomic<uint32_t> expand_count;
+  } MMON_STAT;
+
+  class mmon_subcomponent
+  {
+    public:
+      mmon_subcomponent (const char *subcomp_name)
+	: m_subcomp_name (subcomp_name) {}
+      mmon_subcomponent (const mmon_subcomponent &) = delete;
+      mmon_subcomponent (mmon_subcomponent &) = delete;
+
+      ~mmon_subcomponent () {}
+
+      mmon_subcomponent &operator = (const mmon_subcomponent &) = delete;
+      mmon_subcomponent &operator = (mmon_subcomponent &) = delete;
+
+      /* function */
+      const char *get_name ();
+      void add_cur_stat (uint64_t size);
+      void sub_cur_stat (uint64_t size);
+
+    private:
+      std::string m_subcomp_name;
+      std::atomic<uint64_t> m_cur_stat;
+  };
+
+  class mmon_component
+  {
+    public:
+      mmon_component (const char *comp_name)
+	: m_comp_name (comp_name) {}
+      mmon_component (const mmon_component &) = delete;
+      mmon_component (mmon_component &) = delete;
+
+      ~mmon_component () {}
+
+      mmon_component &operator = (const mmon_component &) = delete;
+      mmon_component &operator = (mmon_component &) = delete;
+
+      /* function */
+      const char *get_name ();
+      void add_stat (uint64_t size, int subcomp_idx, bool init, bool expand);
+      void sub_stat (uint64_t size, int subcomp_idx, bool init);
+      int is_subcomp_exist (const char *subcomp_name);
+      int add_subcomponent (const char *name);
+
+    private:
+      std::string m_comp_name;
+      MMON_STAT m_stat;
+      std::vector<std::unique_ptr<mmon_subcomponent>> m_subcomponent;
+  };
+
+  class mmon_module
+  {
+    public:
+      /* const */
+      /* max index of component or subcomponent is 0x0000FFFF
+       * if some stats have max index of component or subcomponent,
+       * we don't have to increase it */
+      static constexpr int MAX_COMP_IDX = 0x0000FFFF;
+
+    public:
+      mmon_module (const char *module_name, const MMON_COMP_INFO *info);
+      mmon_module () {} /* for dummy */
+      mmon_module (const mmon_module &) = delete;
+      mmon_module (mmon_module &) = delete;
+
+      virtual ~mmon_module () {}
+
+      mmon_module &operator = (const mmon_module &) = delete;
+      mmon_module &operator = (mmon_module &) = delete;
```

Comment:
You may have wanted to do:
```suggestion
      mmon_module &operator = (mmon_module &&) = delete;
```

Reply:
That's right.. Done.